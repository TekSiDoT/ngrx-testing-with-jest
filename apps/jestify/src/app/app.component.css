#addJester {
  position: fixed;
  left: 20px;
  top: 10px;
}

#partyHard {
  position: fixed;
  left: 100px;
  top: 10px;
}

#killParty {
  position: fixed;
  left: 187px;
  top: 10px;
}
