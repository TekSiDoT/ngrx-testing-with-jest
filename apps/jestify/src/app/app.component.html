<div [class.disco]="appFacade.partyHard$ | async"></div>
<button id="addJester" type="button" (click)="appFacade.setJesters(jesters.length + 1)">Add Jester</button>
<button id="partyHard" type="button" (click)="appFacade.partyHard()">Party HARD</button>
<button id="killParty" type="button" (click)="appFacade.killParty()">Kill Party</button>
<jestify-jester *ngFor="let jest of jesters"></jestify-jester>
