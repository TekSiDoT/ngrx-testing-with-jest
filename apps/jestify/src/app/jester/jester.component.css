#jester {
  height: auto;
  position: fixed;
}

.flip {
  transform: scaleX(-1);
}
