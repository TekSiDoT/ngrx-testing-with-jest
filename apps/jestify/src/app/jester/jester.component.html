<img
  id="jester"
  src="../../assets/jester.png"
  [class.flip]="flip$ | async"
  [style.top.px]="(coord$ | async)?.y"
  [style.left.px]="(coord$ | async)?.x"
  [style.width.px]="width"
  [style.filter]="'hue-rotate(' + (color$ | async) + 'deg)' | safe: 'style'"
/>
