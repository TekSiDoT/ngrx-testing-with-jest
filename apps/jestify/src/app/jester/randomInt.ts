// Adapted from https://stackoverflow.com/a/1527820/1792564
export const getRandomInt = (min: number, max: number) => {
  min = Math.ceil(min);
  max = Math.floor(max);
  return Math.floor(Math.random() * (max - min + 1)) + min;
};
