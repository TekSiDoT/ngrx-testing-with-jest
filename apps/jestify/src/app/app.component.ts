import { Component, OnInit, ChangeDetectionStrategy } from '@angular/core';
import { AppFacade } from './+state/app.facade';
import { CookieService } from 'ngx-cookie-service';
import { catchError } from 'rxjs/operators';

export interface Coordinate {
  x: number;
  y: number;
}

@Component({
  selector: 'jestify-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.css'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class AppComponent implements OnInit {
  jesters: number[] = [];
  constructor(public appFacade: AppFacade, private cookieService: CookieService) {}
  ngOnInit() {
    let jesters = 0;
    try {
      if (this.cookieService.check('jesters')) {
        jesters = parseInt(this.cookieService.get('jesters'), 10);
      }
    } catch (e) {
      console.log(e);
    } finally {
      this.appFacade.setJesters(jesters);
    }

    this.appFacade.jesters$.subscribe(cnt => (this.jesters = new Array(cnt)));
  }
}
