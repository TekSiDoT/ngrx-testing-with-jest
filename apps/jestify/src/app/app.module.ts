import { BrowserModule } from '@angular/platform-browser';
import { NgModule } from '@angular/core';

import { AppComponent } from './app.component';
import { NxModule } from '@nrwl/nx';
import { JesterComponent } from './jester/jester.component';
import { StoreModule } from '@ngrx/store';
import { EffectsModule } from '@ngrx/effects';
import { APP_FEATURE_KEY, initialState as appInitialState, appReducer } from './+state/app.reducer';
import { AppEffects } from './+state/app.effects';
import { AppFacade } from './+state/app.facade';
import { StoreDevtoolsModule } from '@ngrx/store-devtools';
import { environment } from '../environments/environment';
import { storeFreeze } from 'ngrx-store-freeze';
import { SafePipeModule } from 'safe-pipe';
import { CookieService } from 'ngx-cookie-service';

@NgModule({
  declarations: [AppComponent, JesterComponent],
  imports: [
    BrowserModule,
    SafePipeModule,
    NxModule.forRoot(),
    StoreModule.forRoot(
      { app: appReducer },
      {
        initialState: { app: appInitialState },
        metaReducers: !environment.production ? [storeFreeze] : []
      }
    ),
    EffectsModule.forRoot([AppEffects]),
    !environment.production ? StoreDevtoolsModule.instrument() : []
  ],
  providers: [AppFacade, CookieService],
  bootstrap: [AppComponent]
})
export class AppModule {}
