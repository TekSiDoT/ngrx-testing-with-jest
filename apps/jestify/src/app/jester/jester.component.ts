import { Component, OnInit, ChangeDetectionStrategy } from '@angular/core';
import { Coordinate } from '../app.component';
import { getRandomInt } from './randomInt';
import { AppFacade } from '../+state/app.facade';
import { tap, filter, map, withLatestFrom, pairwise, scan } from 'rxjs/operators';
import { Observable, interval, timer } from 'rxjs';

@Component({
  selector: 'jestify-jester',
  templateUrl: './jester.component.html',
  styleUrls: ['./jester.component.css'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class JesterComponent implements OnInit {
  width: number = getRandomInt(30, 100);
  coord: Coordinate = {
    x: getRandomInt(0, window.innerWidth),
    y: getRandomInt(0, window.innerHeight)
  };

  color$: Observable<number>;
  coord$: Observable<Coordinate>;
  flip$: Observable<boolean>;

  constructor(public appFacade: AppFacade) {
    this.flip$ = interval(200).pipe(map(() => Math.random() > 0.5));

    this.color$ = interval(200).pipe(
      withLatestFrom(this.appFacade.partyHard$),
      map(isParty => {
        return isParty[1] ? getRandomInt(0, 360) : 0;
      })
    );

    this.coord$ = timer(0, getRandomInt(500, 2000)).pipe(
      map(() => {
        const deltaX = getRandomInt(-10, 10);
        const deltaY = getRandomInt(-10, 10);
        return {
          x: this.coord.x + deltaX < window.innerWidth ? this.coord.x + deltaX : this.coord.x,
          y: this.coord.y + deltaY < window.innerWidth ? this.coord.y + deltaY : this.coord.y
        };
      })
    );
  }

  ngOnInit() {
    this.color$.subscribe();
    this.coord$.subscribe(coord => (this.coord = coord));
  }
}
