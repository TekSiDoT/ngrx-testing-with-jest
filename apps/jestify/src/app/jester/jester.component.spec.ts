import { TestBed, ComponentFixture, async } from '@angular/core/testing';
import { StoreModule, Store } from '@ngrx/store';
import { appReducer, initialState, AppState } from '../+state/app.reducer';
import { EffectsModule } from '@ngrx/effects';
import { AppEffects } from '../+state/app.effects';
import { AppFacade } from '../+state/app.facade';
import { NgModule } from '@angular/core';
import { NxModule } from '@nrwl/nx';
import { readFirst } from '@nrwl/nx/testing';
import { JesterComponent } from './jester.component';
import { CommonModule } from '@angular/common';
import { SafePipeModule } from 'safe-pipe';
import { CookieService } from 'ngx-cookie-service';

interface TestSchema {
  app: AppState;
}

describe('Jester Component', () => {
  let facade: AppFacade;
  let component: JesterComponent;
  let fixture: ComponentFixture<JesterComponent>;

  beforeEach(() => {
    @NgModule({
      imports: [
        CommonModule,
        SafePipeModule,
        StoreModule.forFeature('app', appReducer, { initialState }),
        EffectsModule.forFeature([AppEffects])
      ],
      providers: [AppFacade, CookieService],
      declarations: [JesterComponent]
    })
    class CustomFeatureModule {}

    @NgModule({
      imports: [NxModule.forRoot(), StoreModule.forRoot({}), EffectsModule.forRoot([]), CustomFeatureModule],
      providers: [AppFacade]
    })
    class RootModule {}
    TestBed.configureTestingModule({ imports: [RootModule] }).compileComponents();

    facade = TestBed.get(AppFacade);

    fixture = TestBed.createComponent(JesterComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create the jester component', async done => {
    expect(fixture.debugElement.componentInstance).toBeTruthy();
    done();
  });

  it('should check the initial width range', async done => {
    expect(fixture.componentInstance.width).toBeGreaterThanOrEqual(30);
    expect(fixture.componentInstance.width).toBeLessThanOrEqual(100);
    done();
  });

  it('has correct initial selector states', async done => {
    try {
      const jesters = await readFirst(facade.jesters$);
      expect(jesters).toMatchSnapshot();

      const partyHard = await readFirst(facade.partyHard$);
      expect(partyHard).toMatchSnapshot();

      done();
    } catch (err) {
      done.fail(err);
    }
  });
});
