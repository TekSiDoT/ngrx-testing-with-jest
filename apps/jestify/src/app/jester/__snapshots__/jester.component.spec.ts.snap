// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Jester Component has correct initial selector states 1`] = `1`;

exports[`Jester Component has correct initial selector states 2`] = `false`;
